Improve ArchCondition that checks for transitive dependencies

Pfoerd opened the pull request:
This adds an ArchCondition to the fluent API to check whether there is any matching transitive dependency. This is a much more efficient variant of `#transitivelyDependOnClassesThat` that can be used to detect transitive dependencies in a huge codebase or to classes in large 3rd-party libraries like the Android SDK. It focuses on detecting all *direct* dependencies of the selected classes that are themselves matched or have any transitive dependencies on matched classes. Thus, it doesn't discover all possible dependency paths but stops at the first match to be fast and resource-friendly.

Sample usage:
```
noClasses().that.resideInAPackage(“de.foo.service..”)
    .should().transitivelyDependOnAnyClassesThat.resideInAPackage("android..")
```

Then this Architecture
![image](https://user-images.githubusercontent.com/17569373/172873445-17111662-30e2-4388-912e-840a105cd2bc.png)

would output the following violations:

```
java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'no classes that reside in a package 'de.foo.service..' should transitively depend on any classes that reside in a package ['android..'] was violated (3 times):
Class <de.foo.service.B> transitively depends on <android.I> by [F->E->android.I] in (B.java:0)
Class <de.foo.service.B> transitively depends on <android.I> by [E->android.I] in (B.java:0)
Class <de.foo.service.C> transitively depends on <android.H> by [D->android.H] in (C.java:0)
```

Resolves: #780 
Resolves: #826

Pfoerd commented:
Hi @codecholeric. Do you have time for some feedback on the PR? And maybe you could approve the execution of workflows? Thanks!

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+        super("transitively depend on any classes that " + conditionPredicate.getDescription());
+
+        this.conditionPredicate = checkNotNull(conditionPredicate);
+        this.transitiveDependencyPath = new JavaClassTransitiveDependencyPath();
```
Sure! To be honest, I haven't developed productively in Java for a long time, I've mainly been doing Kotlin for the last few years. Looks like I'm a bit rusty :D

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    @Override
+    public void check(JavaClass item, ConditionEvents events) {
```
Sure

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+         */
+        List<JavaClass> findFirstPathToTransitiveDependency(JavaClass clazz) {
+            LinkedList<JavaClass> transitivePath = new LinkedList<>();
+            addDependenciesToPathFrom(clazz, transitivePath, new HashSet<>());
```
Well spotted! You're absolutely right, the repetitive `contains` call of the unsorted collection inside the recursion is obviously very expensive

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+         * @return the first dependency path to a matching class or empty if there is none
+         */
+        List<JavaClass> findFirstPathToTransitiveDependency(JavaClass clazz) {
+            LinkedList<JavaClass> transitivePath = new LinkedList<>();
+            addDependenciesToPathFrom(clazz, transitivePath, new HashSet<>());
+            return Collections.unmodifiableList(transitivePath);
```
Sure!

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+    private static Set<JavaClass> getDirectDependencies(JavaClass item) {
```
Sure!

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+        return directDependencies;
+    }
+
+    private static class TransitivePathFoundEvent implements ConditionEvent {
```
Sure.

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+            this.selected = selected;
+
+            String messageStr =
+                    String.format("Class <%s> accesses <%s>", selected.getFullName(), transitivePath.get(0).getFullName());
```
You're totally right, the wording here was not very precise

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+        public TransitivePathFoundEvent(JavaClass selected, List<JavaClass> transitivePath) {
+            this.selected = selected;
+
+            String messageStr =
+                    String.format("Class <%s> accesses <%s>", selected.getFullName(), transitivePath.get(0).getFullName());
+
+            if (transitivePath.size() > 1) {
+                messageStr += " which transitively accesses e.g. ";
+                StringJoiner joiner = new StringJoiner("> <- <", "<", ">");
+                for (JavaClass dependencyTarget : Lists.reverse(transitivePath)) {
+                    joiner.add(dependencyTarget.getFullName());
+                }
+                messageStr += joiner.toString();
```
Thank you for your comprehensive feedback, i like the new message format! 

What about adding a "transitively" before "depends on"?
```
String messageStr =
        String.format("Class <%s> transitively depends on <%s>", selected.getFullName(), getLast(transitivePath).getFullName());
```

hankem replied:
I like the additional "transitively" if `transitivePath.size() > 1`. 😉

Pfoerd replied:
Good point! I add a condition!

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,136 @@
+        for (JavaClass dependency : getDirectDependencies(item)) {
+            if (!allClasses.contains(dependency)) {
+                List<JavaClass> dependencyPath = transitiveDependencyPath.findFirstPathToTransitiveDependency(dependency);
+                if (!dependencyPath.isEmpty()) {
+                    events.add(new TransitivePathFoundEvent(item, dependencyPath));
+                }
```
Okay so the solution for that would be to add something like 
```
if (!dependencyPath.isEmpty()) {
    events.add(newTransitivePathFoundEvent(javaClass, dependencyPath));
} else {
    events.add(noTransitivePathFoundEvent(javaClass));
}
...
private static ConditionEvent noTransitivePathFoundEvent(JavaClass selected) {
    return SimpleConditionEvent.violated(selected, String.format("Class <%s> not transitively depends on any matching class", selected.getFullName());
}
```
right? Can do that np, what do you think?

Pfoerd commented on archunit/src/main/java/com/tngtech/archunit/lang/syntax/elements/ClassesShould.java:
```diff
@@ -1015,6 +1015,48 @@ public interface ClassesShould {
+     * It focuses on detecting all <strong>direct</strong> dependencies of the selected classes that are themselves matched or have any
+     * transitive dependencies on matched classes. Thus, it doesn't discover all possible dependency paths but stops at the first match to be fast and
+     * resource-friendly.<br>
+     * NOTE: This usually makes more sense the negated way, e.g.
```
okay see my [comment above](https://github.com/TNG/ArchUnit/pull/878/files#r904886323) regarding the "negated" way

Pfoerd commented on archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/ShouldClassesThatTest.java:
```diff
@@ -393,7 +393,8 @@ public void containAnyMembersThat(ClassesThat<ClassesShouldConjunction> noClasse
                         Data_of_containAnyMembersThat.OkayTarget.class, Data_of_containAnyMembersThat.ViolatingTarget.class);
 
-        assertThatTypes(classes).matchInAnyOrder(Data_of_containAnyMembersThat.ViolatingOrigin.class, Data_of_containAnyMembersThat.ViolatingTarget.class);
+        assertThatTypes(classes).matchInAnyOrder(Data_of_containAnyMembersThat.ViolatingOrigin.class,
```
ah sorry, it was the last autoformat before pushing ;) .

Pfoerd commented on archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/ShouldClassesThatTest.java:
```diff
@@ -1708,49 +1714,123 @@ public void onlyDependOnClassesThat_reports_all_dependencies() {
+        }
+
+        @SuppressWarnings("unused")
+        static class TestClass2 {
```
Anyone who thinks naming is difficult in computer science has never had to choose names in unit tests :D ... The thing here is that the same class structure is used in the tests for the "old" transitive condition and there it causes violations. Thus, i would prefer to keep the naming a little more generic or we would have to double the class structure. What do you think?

codecholeric replied:
As mentioned below, if we remove the old API then we should be able to rename it, right?

Pfoerd commented on archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/ShouldClassesThatTest.java:
```diff
@@ -1708,49 +1714,123 @@ public void onlyDependOnClassesThat_reports_all_dependencies() {
+        static class TestClass1 {
             DirectlyDependentClass1 directDependency1;
             DirectlyDependentClass2 directDependency2;
+            TransitivelyDependentClass3 directDependency3;
```
thx, I fix the matcher. I think I could use "belongToAnyOf" and pass the three types?

codecholeric replied:
Yeah, true, that might be the simplest :+1:

Pfoerd commented on archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/ShouldClassesThatTest.java:
```diff
@@ -1708,49 +1714,123 @@ public void onlyDependOnClassesThat_reports_all_dependencies() {
+                ? noClassesShould.transitivelyDependOnAnyClassesThat(have(simpleNameStartingWith("TransitivelyDependentClass")))
+                : noClassesShould.transitivelyDependOnAnyClassesThat().haveSimpleNameStartingWith("TransitivelyDependentClass");
+
+        assertThatRule(rule).checking(classes)
```
i added another layer of transitivity and changed naming to match the new structure. please have a look

codecholeric replied:
Looks good :+1:

Pfoerd commented:
Hi, no regrets here, i really appreciate comprehensive feedback :) ! I was able to correct most of your comments right away, see my comments with the rest

I was struggling with the namings around `transitivelyDependOnAnyClasses` too. I like your idea to just replace the old implementation (at least because it solves my naming issue here 😆 ).

hankem commented:
I don't mind replacing the old `transitivelyDependOnClassesThat` at all. Thanks!

Pfoerd commented:
> I don't mind replacing the old `transitivelyDependOnClassesThat` at all. Thanks!

@codecholeric should I then remove the "old" implementation, keep the original fluent API `transitivelyDependOnClasses` and just replace the implementation?

codecholeric commented:
> > I don't mind replacing the old `transitivelyDependOnClassesThat` at all. Thanks!
> 
> @codecholeric should I then remove the "old" implementation, keep the original fluent API `transitivelyDependOnClasses` and just replace the implementation?

Yes, I think this makes sense! Because it will make the API easier to understand and if we can't think of any use case where the other implementation is more appropriate, then we should just remove it. It will be a breaking change though, so perfect opportunity now for ArchUnit 1.0 :wink: 
As a side effect it will get rid of your test naming problem :wink: Then you can freely rename `TestClass2`, right? :grin:

codecholeric commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public void check(JavaClass javaClass, ConditionEvents events) {
+        boolean hasTransitiveDependency = false;
+        for (JavaClass dependency : getDirectDependencies(javaClass)) {
```
what about making the method `getRelevantDirectDependencies(..)` or sth. like that and already filtering out `allClasses`? That would save one level of nesting here :slightly_smiling_face:

codecholeric commented on archunit/src/main/java/com/tngtech/archunit/lang/conditions/AnyTransitiveDependencyCondition.java:
```diff
@@ -0,0 +1,112 @@
+
+    private static ConditionEvent noTransitivePathFoundEvent(JavaClass selected) {
+        return SimpleConditionEvent.violated(selected,
+                String.format("Class <%s> not transitively depends on any matching class", selected.getFullName()));
```
should it be "does not transitively depend on..."?

codecholeric commented:
Yes, looks good now :partying_face: I did some minor changes, e.g. there was a commit in the other PR to more conveniently create condition messages that I cherry-picked in here to merge with this one instead. Other than that very nice and ready to merge :smile:

codecholeric commented:
Ah, but I just saw that I can't push my changes, `permission denied` :disappointed: Would you mind giving me access? I also rebased on `main` to solve that issue and added the one missing copyright note...

Pfoerd commented:
> Ah, but I just saw that I can't push my changes, `permission denied` 😞 Would you mind giving me access? I also rebased on `main` to solve that issue and added the one missing copyright note...

done!

Pfoerd commented:
@codecholeric thx for your support! Looking forward to see this in ArchUnit 1.0 ❤️

codecholeric commented:
Well, thanks for your contribution :smile: Without it I would certainly not have gotten around to improve this now...
